### PR TITLE
feat: implement role subscription fields

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

https://github.com/discord/discord-api-docs/pull/5724

The new integration type is untested due to region locks (:/), though I don't see why it wouldn't work as the integration structure wasn't changed.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/emoji.py [hunk 1]
```diff
@@ -145,6 +145,9 @@ def roles(self) -> List[Role]:
         """List[:class:`Role`]: A :class:`list` of roles that are allowed to use this emoji.
 
         If roles is empty, the emoji is unrestricted.
+
+        Emojis with :attr:`subscription roles <RoleTags.integration_id>` are considered premium emoji,
```
Comment (Victorsitou) [inline disnake/emoji.py:149]:
```suggestion
        Emojis with :attr:`subscription roles <RoleTags.integration_id>` are considered premium emojis,
```

File: disnake/guild.py [hunk 2]
```diff
@@ -3227,6 +3234,9 @@ async def create_custom_emoji(
             "2", "150"
             "3", "250"
 
+        Emojis with :attr:`subscription roles <RoleTags.integration_id>` are considered premium emoji,
```
Comment (Victorsitou) [inline disnake/guild.py:3237]: This is missing a note saying that you can't mix subscription roles and non-subscription roles when creating a custom emoji, (here and Role.edit)

**Review by Victorsitou — COMMENTED** [review]
Tested it, and it works fine, left some docs reviews c:

File: disnake/guild.py (hunk 2 above)
Comment (shiftinv) [reply disnake/guild.py]: clarified it a bit and added more details in https://github.com/DisnakeDev/disnake/pull/904/commits/07bf0b7569d07547982f0cfa7748bd83a77ecc95

File: disnake/guild.py [hunk 3]
```diff
@@ -847,7 +850,11 @@ def public_updates_channel(self) -> Optional[TextChannel]:
 
     @property
     def emoji_limit(self) -> int:
-        """:class:`int`: The maximum number of emoji slots this guild has."""
+        """:class:`int`: The maximum number of emoji slots this guild has.
+
+        Premium emojis (i.e. those associated with subscription roles) count towards a
+        separate limit of 25.
```
Comment (onerandomusername) [inline disnake/guild.py:856]: thoughts on a possible `premium_emoji_limit` property of 25?

not sure if that's a good idea or not tbh

File: docs/api.rst [hunk 4]
```diff
@@ -1764,6 +1764,11 @@ of :class:`enum.Enum`.
         The system message denoting that an auto moderation action was executed.
 
         .. versionadded:: 2.5
+    .. attribute:: role_subscription_purchase
+
+        The system message denoting that a role subscription was purchased.
+
+        .. versionadded:: 2.8
```
Comment (onerandomusername) [inline docs/api.rst:1771]: 2.9

File: disnake/role.py [hunk 5]
```diff
@@ -46,24 +46,37 @@ class RoleTags:
         The bot's user ID that manages this role.
     integration_id: Optional[:class:`int`]
         The integration ID that manages the role.
+
+        Roles with this ID matching the guild's ``guild_subscription`` integration
```
Comment (onerandomusername) [inline disnake/role.py:50]: thoughts on a shortcut for this? a property on the role perhaps? 

otoh we can implement that later after the api is more finalised and if there's demand

File: disnake/integrations.py [hunk 6]
```diff
@@ -119,7 +119,7 @@ class Integration(PartialIntegration):
         Whether the integration is currently enabled.
     account: :class:`IntegrationAccount`
         The account linked to this integration.
-    user: :class:`User`
+    user: Optional[:class:`User`]
```
Comment (onerandomusername) [inline disnake/integrations.py:122]: versionchanged?

**Review by onerandomusername — CHANGES_REQUESTED** [review]
tested as much as was possible, looks good other than these two comments

File: disnake/guild.py (hunk 3 above)
Comment (shiftinv) [reply disnake/guild.py:856]: if a nitro-dependent number of premium emoji slots gets added, that would make sense, yeah. But I don't think we need one for a constant value just yet.

File: docs/api.rst (hunk 4 above)
Comment (shiftinv) [reply docs/api.rst]: https://github.com/DisnakeDev/disnake/pull/904/commits/0b4579ac72edaa3b50e8dea243263e83dae2128e

File: disnake/integrations.py (hunk 6 above)
Comment (shiftinv) [reply disnake/integrations.py:122]: The documentation was outdated, it's been optional for a while now.

File: disnake/role.py (hunk 5 above)
Comment (shiftinv) [reply disnake/role.py:50]: Would be neat, but afaik there's no easy way of obtaining the `guild_subscription` integration ID without fetching integrations (which requires `manage_guild` perms).

**Review by onerandomusername — APPROVED** [review]

**Pull request merged**